vespa-feed-client/src/main/java/ai/vespa/feed/client/JsonFeeder.java

```diff
@@ -68,45 +99,61 @@ private JsonStreamFeeder(FeedClient client, OperationParameters protoParameters)
      * </pre>
      * Note that {@code "id"} is an alias for the document put operation.
      */
-    public void feed(InputStream jsonStream) throws IOException {
-        feed(jsonStream, 1 << 26, false);
+    public CompletableFuture<Void> feedMany(InputStream jsonStream, ResultCallback resultCallback) {
+        return feedMany(jsonStream, 1 << 26, resultCallback);
     }
 
-    BenchmarkResult benchmark(InputStream jsonStream) throws IOException {
-        return feed(jsonStream, 1 << 26, true).get();
-    }
-
-    Optional<BenchmarkResult> feed(InputStream jsonStream, int size, boolean benchmark) throws IOException {
+    CompletableFuture<Void> feedMany(InputStream jsonStream, int size, ResultCallback resultCallback) {
         RingBufferStream buffer = new RingBufferStream(jsonStream, size);
-        buffer.expect(JsonToken.START_ARRAY);
-        AtomicInteger okCount = new AtomicInteger();
-        AtomicInteger failedCount = new AtomicInteger();
-        long startTime = System.nanoTime();
+        CompletableFuture<Void> overallResult = new CompletableFuture<>();
         CompletableFuture<Result> result;
-        AtomicReference<Throwable> thrown = new AtomicReference<>();
-        while ((result = buffer.next()) != null) {
-            result.whenComplete((r, t) -> {
-                if (t != null) {
-                    failedCount.incrementAndGet();
-                    if (!benchmark) thrown.set(t);
-                } else
-                    okCount.incrementAndGet();
-            });
-            if (thrown.get() != null)
-                sneakyThrow(thrown.get());
+        AtomicInteger pending = new AtomicInteger(1); // The below dispatch loop itself is counted as a single pending operation
+        AtomicBoolean finalCallbackInvoked = new AtomicBoolean();
+        try {
+            while ((result = buffer.next()) != null) {
+                pending.incrementAndGet();
+                result.whenComplete((r, t) -> {
+                    if (!finalCallbackInvoked.get()) {
+                        resultExecutor.execute(() -> resultCallback.onNextResult(r, t));
+                    }
+                    if (pending.decrementAndGet() == 0 && finalCallbackInvoked.compareAndSet(false, true)) {
+                        resultExecutor.execute(() -> {
+                            resultCallback.onComplete();
+                            overallResult.complete(null);
+                        });
+                    }
+                });
```

Comment:
```suggestion
                result.whenCompleteAsync((r, t) -> {
                    if (!finalCallbackInvoked.get()) {
                        resultCallback.onNextResult(r, t);
                    }
                    if (pending.decrementAndGet() == 0 && finalCallbackInvoked.compareAndSet(false, true)) {
                        resultCallback.onComplete();
                        overallResult.complete(null);
                    }
                },
                resultExecutor);
```
Let result executor do more work, for less contention?